src/borg/_hashindex.c

```diff
@@ -351,23 +359,168 @@ hashindex_read(PyObject *file_py, int permit_compact)
     }
 
     index->num_entries = _le32toh(header->num_entries);
     index->num_buckets = _le32toh(header->num_buckets);
+    index->num_empty = -1;  // unknown, needs counting
     index->key_size = header->key_size;
     index->value_size = header->value_size;
+
+fail_release_header_buffer:
+    PyBuffer_Release(&header_buffer);
+fail_free_index:
+    if(PyErr_Occurred()) {
+        free(index);
+        index = NULL;
+    }
+fail_decref_header:
+    Py_DECREF(header_bytes);
+fail:
+    return index;
+}
+
+HashIndex *
+read_hashheader(PyObject *file_py)
+{
+    Py_ssize_t bytes_read, length, buckets_length;
+    Py_buffer header_buffer;
+    PyObject *header_bytes, *length_object, *tmp;
+    HashIndex *index = NULL;
+    HashHeader *header;
+
+    header_bytes = PyObject_CallMethod(file_py, "read", "n", (Py_ssize_t)sizeof(*header));
+    if(!header_bytes) {
+        assert(PyErr_Occurred());
+        goto fail;
+    }
+
+    bytes_read = PyBytes_Size(header_bytes);
+    if(PyErr_Occurred()) {
+        /* TypeError, not a bytes() object */
+        goto fail_decref_header;
+    }
+    if(bytes_read != sizeof(*header)) {
+        /* Truncated file */
+        /* Note: %zd is the format for Py_ssize_t, %zu is for size_t */
+        PyErr_Format(PyExc_ValueError, "Could not read header (expected %zu, but read %zd bytes)",
+                     sizeof(*header), bytes_read);
+        goto fail_decref_header;
+    }
+
+    /*
+     * Hash the header
+     * If the header is corrupted this bails before doing something stupid (like allocating 3.8 TB of memory)
+     */
+    tmp = PyObject_CallMethod(file_py, "hash_part", "s", "HashHeader");
+    Py_XDECREF(tmp);
+    if(PyErr_Occurred()) {
+        if(PyErr_ExceptionMatches(PyExc_AttributeError)) {
+            /* Be able to work with regular file objects which do not have a hash_part method. */
+            PyErr_Clear();
+        } else {
+            goto fail_decref_header;
+        }
+    }
+
+    /* Find length of file */
+    length_object = PyObject_CallMethod(file_py, "seek", "ni", (Py_ssize_t)0, SEEK_END);
+    if(PyErr_Occurred()) {
+        goto fail_decref_header;
+    }
+    length = PyNumber_AsSsize_t(length_object, PyExc_OverflowError);
+    Py_DECREF(length_object);
+    if(PyErr_Occurred()) {
+        /* This shouldn't generally happen; but can if seek() returns something that's not a number */
+        goto fail_decref_header;
+    }
+
+    tmp = PyObject_CallMethod(file_py, "seek", "ni", (Py_ssize_t)sizeof(*header), SEEK_SET);
+    Py_XDECREF(tmp);
+    if(PyErr_Occurred()) {
+        goto fail_decref_header;
+    }
+
+    /* Set up the in-memory header */
+    if(!(index = malloc(sizeof(HashIndex)))) {
+        PyErr_NoMemory();
+        goto fail_decref_header;
+    }
+
+    PyObject_GetBuffer(header_bytes, &header_buffer, PyBUF_SIMPLE);
+    if(PyErr_Occurred()) {
+        goto fail_free_index;
+    }
+
+    header = (HashHeader*) header_buffer.buf;
```

Comment:
Hm, why does this not cause issues with unaligned memory accesses on non-x86 architectures?

I mean it seems like it does not cause such issues, because the old code does the same and I have it running on the same platform I found [the lack of this mpd commit to be an issue](https://github.com/MusicPlayerDaemon/MPD/commit/fb8d8242abdbb0001012c5b91b4e2ca69553b166), but it leaves me curious.

Are we just lucky that `header_buffer.buf` is actually aligned to an 8-byte boundary and the header magic is a multiple of 8?

Reply:
Well I'ld hope python is clever enough to allocate its stuff in a well-aligned way.